src/cardano_fmt_plug.c

```diff
@@ -0,0 +1,244 @@
+ * and it is hereby released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_cardano;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_cardano);
+#else
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+#define OMP_SCALE 1
+
+#include "arch.h"
+#include "misc.h"
+#include "common.h"
+#include "formats.h"
+#include "params.h"
+#include "options.h"
+#include "jumbo.h"
+#include "cardano_common.h"
+#include "blake2.h"
+#include "pbkdf2_hmac_sha512.h"
+#include "chacha.h"
+#include "ed25519.h"
+
+#define FORMAT_NAME         "Cardano Encrypted 128-byte Secret Key (a.k.a XPrv)"
+#define FORMAT_LABEL        "cardano"
+#ifdef SIMD_COEF_64
+#define ALGORITHM_NAME      "PBKDF2-SHA512/BLAKE2b/ChaCha20 " SHA512_ALGORITHM_NAME
+#else
+#if ARCH_BITS >= 64
+#define ALGORITHM_NAME      "PBKDF2-SHA512/BLAKE2b/ChaCha20  64/" ARCH_BITS_STR
+#else
+#define ALGORITHM_NAME      "PBKDF2-SHA512/BLAKE2b/ChaCha20  32/" ARCH_BITS_STR
+#endif
+#endif
+#define BENCHMARK_COMMENT   ""
+#define BENCHMARK_LENGTH    0x107
+#define BINARY_SIZE         0
+#define BINARY_ALIGN        1
+#define SALT_SIZE           sizeof(struct custom_salt)
+#define SALT_ALIGN          sizeof(uint32_t)
+#ifdef SIMD_COEF_64
+#define MIN_KEYS_PER_CRYPT  SSE_GROUP_SZ_SHA512
+#define MAX_KEYS_PER_CRYPT  (SSE_GROUP_SZ_SHA512 * 4)
+#else
+#define MIN_KEYS_PER_CRYPT  1
+#define MAX_KEYS_PER_CRYPT  4
+#endif
+#if !defined(PLAINTEXT_LENGTH) || PLAINTEXT_LENGTH > 15
+#undef PLAINTEXT_LENGTH
+#define PLAINTEXT_LENGTH    15
+#endif
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int any_cracked, *cracked;
+static size_t cracked_size;
+
+static struct custom_salt *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+	omp_autotune(self, OMP_SCALE);
+
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	any_cracked = 0;
+	cracked = mem_calloc(cracked_size, 1);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void cardano_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(*saved_key));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+
+	if (any_cracked) {
+		memset(cracked, 0, cracked_size);
+		any_cracked = 0;
+	}
+	// KDF2 params
+	static const int kdf_rounds = 15000;
+	static const char kdf_salt[] = "encrypted wallet salt";
+	static int kdf_salt_len = sizeof(kdf_salt);
+
+	// ChaCha20 params
+	static const int chacha_rounds = 20;
+
+#ifdef _OPENMP
+	#pragma omp parallel for
+#endif
```

Comment:
We normally don't indent `#pragma` directives - just keep them on the leftmost column. Ditto for the other one below.

Reply:
Fixed